pkg/ipcache/gc.go

```diff
@@ -41,15 +43,22 @@ func newAsyncPrefixReleaser(parent prefixReleaser, interval time.Duration) *asyn
 		Name:        "ipcache-identity-gc",
 		MinInterval: interval,
 		TriggerFunc: func(reasons []string) {
-			// TODO: Structure the code to pass context down
-			//       from the Daemon.
-			ctx, cancel := context.WithTimeout(
-				context.TODO(),
+			innerCtx, cancel := context.WithTimeout(
+				ctx,
 				option.Config.KVstoreConnectivityTimeout)
 			defer cancel()
-			result.run(ctx, reasons...)
+			result.run(innerCtx, reasons...)
 		},
 	})
+	go func() {
+		<-ctx.Done()
+		// Prevent callers from injecting new prefixes and block on
+		// the closed trigger
+		result.Lock()
+		result.Trigger.Shutdown()
+		close(result.closeChan)
+		result.Unlock()
+	}()
```

Comment:
Is this goroutine simply running to check if the daemon context is done, and then trigger the shutdown in ipcache?
Have you looked into introducing a `Close`/`Shutdown` function in ipcache that'll be invoked during daemon shutdown - https://github.com/cilium/cilium/blob/master/daemon/cmd/daemon.go#L1363? 
This approach has the benefit that we have some kind of guarantee that the ipcache shutdown has completed. However, we need to ensure that the shutdown completes within a reasonable time.

Reply:
Broadly yeah that's what it's doing. The specifics here are relating to the way that events are enqueued below though, because if we *only* call `result.Trigger.Shutdown()` but then other packages continue to enqueue CIDRs into the asyncPrefixReleaser, then those calls will hang when they attempt to do the Trigger(). So this is wrapping that to also have this extra channel to avoid hanging on the trigger.

I hadn't thought too hard about an ipcache.Close/Shutdown function yet, that sounds like a reasonable alternative solution to this part of the problem. I'll look into that.

Reply:
I looked a little closer, unfortunately it's not quite this simple. The reason that this doesn't confirm that the shutdown is complete is because the `pkg/trigger:Trigger.Shutdown()` method only closes a channel, so that function itself does not ensure the logic underneath completes. We'd need to make pkg/trigger more complicated to ensure that occurs synchronously. Hmm.

Reply:
Maybe I can just go ahead with this structure anyway for now, then we can always further close that gap in future. I think it'll likely help the ongoing modularization effort if the IPCache has a shutdown function like this to handle cleanup.

Reply:
>The reason that this doesn't confirm that the shutdown is complete is because the pkg/trigger:Trigger.Shutdown() method only closes a channel

Can we still move the logic to a shutdown function, and invoke that from daemon shutdown? In future, we can make the ipcache/gc.shutdown synchronous, and thereby guarantee a total shutdown (within bounded time). 

```
:
result.Trigger.Shutdown()
close(result.closeChan)
```

Reply:
Almost yes - it turns out that the `Trigger` has a `ShutdownFunc()` which is called once the trigger completes its current task. Using that we can close the loop by having a shutdown function called from the daemon which calls Trigger.Shutdown(), which asynchronously stops any current trigger run then calls the Trigger.ShutdownFunc. From there we can `close(result.closeChan)`. Then, back in the overall ipcache shutdown func which called the `Trigger.Shutdown()` it can block on the close of `resultChan`, thereby ensuring that the trigger func is no longer executing any subsequent logic.

Reply:
I've implemented this suggestion now. I'll mark the PR is draft until I see confirmation that I didn't break the tests then re-mark for review.

Reply:
Thanks for the feedback!